src/scripts/bad_medgen_xrefs_update_analyze_diff.py

```diff
@@ -0,0 +1,87 @@
+"""MedGen conflict updates QC analysis
```

Comment:
### Analysis of `grep -v` results
TLDR: I think this method did not work out.
Probably should instead do (a) SPARQL update, (b), manual/partial editing based on my analysis results (c) remove all xrefs and add from scratch w/ new ingest as originally planned.

---

I analyzed the 5 sheets in [Megan's excel](https://docs.google.com/spreadsheets/d/13CO5yAAhMZD3BhAK8S1GfLF6ZNfDZdcW/edit#gid=61010183). For each sheet, I was able to identify which xrefs to MedGen/UMLS IDs needed to be removed, totaling 2,934.

However, after doing `grep -v` on `mondo-edit.obo` on these IDs, there were more than that number of lines removed, totaling 3,125. Here's a breakdown of my analysis ([report-qc-medgen-conflicts-update-diff.tsv.zip](https://github.com/monarch-initiative/mondo/files/12353911/report-qc-medgen-conflicts-update-diff.tsv.zip)).

```
n_lines_removed 3125
- n_non_xrefs_removals     6
  - synonyms: 4
  - comments: 1
  - relationships: 1
- n_xrefs_removed     3119
  - n_xrefs_removed_cuz_umls_was_a_source   60
  - n_umls_removals 3059
    - n_single_cui_removed_from_xrefs_n_times__1   2819 (1x2819=2819)
    - n_single_cui_removed_from_xrefs_n_times__2   106 (2x106=212)
    - n_single_cui_removed_from_xrefs_n_times__3   8 (3x8=24)
    - n_single_cui_removed_from_xrefs_n_times__4   1 (1x4=4)
```

For each indent line you see here, the indented level below it contains all the different cases that add up to the total on that line.
Out of 3,125 removals, 6 of those were mistakes; they were not xrefs (see below for detail).
3,119 xrefs were removed. However, 60 of those were mistakes; they weren't UMLS/MedGen xrefs. They got removed by `grep` because the MedGen ID was one of the xref sources.
3,059 UMLS/MedGen xrefs were removed. However, sometimes multiple xrefs got removed for a single ID. For 2,819 IDs, that same number of xrefs got removed. 106 IDs were removed 2 times, 8 were removed 3 times, and 1 was removed 4 times _Why this is hard to fix using this `grep -v` approach; hard to explain succinctly_: This is not what Megan/we wanted in most cases. There is only 1 sheet, I believe, where Megan lists out MedGen IDs that no longer exist and that we should remove. But the other sheets talk about specific xrefs on specific Mondo classes. So this number of 3,059 is a problem. We were looking to remove around 2,934 IDs (total number of unique IDs in Megan's excel). It's possible because of that 1 sheet just mentioned, that number could be higher than 2,934. I'd need to re-run this removal pipeline on that sheet alone, but I have a feeling that this difference of excess xrefs removed is likely not coming from the CUIs in that 1 sheet. The crux of it is that even if we solve the issue of non-umls xrefs being removed by the `grep -v` approach, there is still a big problem of too many xrefs being removed here.

<details><summary>Non-xref removals (4 synonyms, 1 comment, 1 relationship)</summary>
<p>

synonym: ""Ullrich disease"" RELATED [DOID:0050558, https://www.ncbi.nlm.nih.gov/gtr/conditions/C0410179/]
relationship: excluded_subClassOf MONDO:0016697 {source=""UMLS:CN201941""} ! low grade ependymoma
comment: Other ontologies classify peripheral neuroepithelioma as a seprate class but NCIT treats these as synonyms and we aim to follow NCIT here. (See HP:0006717, SNOMED:254764001, the ""included"" disease concept on OMIM:612219, and UMLS:C3489398).
synonym: ""polysyndactyly"" EXACT [SCTID:84598000, UMLS:C0265553]
synonym: ""thrombophilia, hereditary"" EXACT [MESH:C540694, UMLS:C2584620]
synonym: ""triple positive breast cancer"" EXACT [UMLS:C5243562]

</p>
</details>

Reply:
I see your point.. It is unfortunate that this has become such a complex issue. Unfortuntely - I don't know how to help you here, you will just have to figure it out somehow! Use whatever means you want!

For the non-xref matches, I would just delete them in isolation (not -v, just the xrefs inside, manually).

> Probably should instead do (a) SPARQL update, (b), manual/partial editing based on my analysis results (c) remove all xrefs and add from scratch w/ new ingest as originally planned.

(a) may be much more cumbersome than you think, because of axiom annotations
(b) yes
(c) I don't think this is easily possible as we may have valid UMLS cross references that are not in Medgen at all..

Reply:
No problem. I need to / should be able to figure out implementations / workarounds. 

I was keen on 'a' SPARQL UPDATE, and aware of the axiom annotations. I was thinking that they would all be the same "shape", so that my query would only need 2 blocks: (i) to remove the xref, and (ii) to remove the axiom annotation. But I think I'll trust your instinct that (a) if I'm correct, implementing this might be more complicated than it seems, or (b) I'm not correct, and there are different shapes / types of axiom annotations, requiring even more blocks in my query.

I hear you on 'c'. I didn't think of that.

So yes, I think I'll go with 'b', manual manipulations, and figure it out as I go.

Reply:
Strategy:
**n_single_cui_removed_from_xrefs_n_times__2+** (~125): Joe will add back xrefs that were accidentally removed.
**n_xrefs_removed_cuz_umls_was_a_source (60)**: Joe will remove manually.
**n_non_xrefs_removals (6)**: Nico removing manually.